tidy3d/components/time_modulation.py

```diff
@@ -0,0 +1,244 @@
+    amplitude: Union[float, SpatialDataArray] = pd.Field(
+        1,
+        title="Amplitude of modulation in space",
+        description="Amplitude of modulation that can vary spatially. "
+        "It takes the unit of whatever is being modulated.",
+    )
+
+    phase: Union[float, SpatialDataArray] = pd.Field(
+        0,
+        title="Phase of modulation in space",
+        description="Phase of modulation that can vary spatially.",
+        units=RADIAN,
+    )
+
+    interp_method: InterpMethod = pd.Field(
+        "nearest",
+        title="Interpolation method",
+        description="Method of interpolation to use to obtain values at spatial locations on the Yee grids.",
+    )
+
+    @pd.validator("amplitude", always=True)
+    def _real_amplitude(cls, val):
+        """Assert that the amplitude is real."""
+        if np.iscomplexobj(val):
+            raise ValidationError("'amplitude' must be real.")
+        return val
+
+    @pd.validator("phase", always=True)
+    def _real_phase(cls, val):
+        """Assert that the phase is real."""
+        if np.iscomplexobj(val):
+            raise ValidationError("'phase' must be real.")
+        return val
+
+    @cached_property
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+        return np.max(abs(np.array(self.amplitude)))
+
+
+SpaceModulationType = Union[SpaceModulation]
+
+
+class SpaceTimeModulation(Tidy3dBaseModel):
+    """Space-time modulation applied to a medium, adding
+    on top of the time-independent permittivity of the medium.
+
+
+    Note
+    ----
+    The space-time modulation must be separable in space and time.
+    e.g. when applied to permittivity,
+
+    .. math::
+
+        \\delta \\epsilon(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+    """
+
+    space_modulation: SpaceModulationType = pd.Field(
+        SpaceModulation(),
+        title="Space modulation",
+        description="Space modulation from the separable SpaceTimeModulation.",
+        # discriminator=TYPE_TAG_STR,
+    )
+
+    time_modulation: TimeModulationType = pd.Field(
+        ...,
+        title="Time modulation",
+        description="Time modulation from the separable SpaceTimeModulation.",
+        # discriminator=TYPE_TAG_STR,
+    )
+
+    @cached_property
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+        return self.time_modulation.max_modulation * self.space_modulation.max_modulation
+
+    @cached_property
+    def negligible_modulation(self) -> bool:
+        """whether the modulation is weak enough to be regarded as zero."""
+        # if isclose(np.diff(time_modulation.range), 0) and
+        if isclose(self.max_modulation, 0):
+            return True
+        return False
+
+
+class ModulationSpec(Tidy3dBaseModel):
+    """Specification adding space-time modulation to the non-dispersive part of medium
+    including epsilon at infinity and electric conductivity.
```

Comment:
I think here and in the permittivity, we should be more clear that this is relative permittivity.  Also, maybe "epsilon at infinity" is too vague (what is epsilon? what=infinity?), maybe for this spec we can just refer to it as `relative permittivity at infinite frequency` if that is the most specific?